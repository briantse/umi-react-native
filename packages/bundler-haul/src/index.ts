import { dirname, join } from 'path';
import { fork } from 'child_process';
import { IApi } from '@umijs/types';
import { yargs } from '@umijs/utils';
import { name } from '../package.json';
import generateFiles from './generateFiles';
import babelConfigTpl from './babelConfigTpl';
import haulConfigTpl from './haulConfigTpl';
import clean from './asyncClean';

export interface IHaulStartOptions {
  port?: number;
  dev: boolean;
  interactive?: boolean;
  minify?: boolean;
  tempDir?: string;
  config: string;
  eager: string;
  maxWorkers?: number;
  skipHostCheck: boolean;
}

interface ICommand {
  name: string;
  alias?: string;
  description?: string;
  details?: string;
  fn: {
    <T>({ args }: { args: yargs.Arguments<T> }): void;
  };
}

export default (api: IApi) => {
  const {
    utils: { portfinder, semver, Mustache, lodash, resolve, winPath },
    paths: { absTmpPath },
  } = api;
  async function handler({ args }: { args: yargs.Arguments<IHaulStartOptions> }): Promise<void> {
    const defaultPort = process.env.PORT || args?.port;
    const port = await portfinder.getPortPromise({
      port: defaultPort ? parseInt(String(defaultPort), 10) : 8081,
    });
    const argv: string[] = ['start', '--config', join(absTmpPath || '', 'haul.config.js')];

    if (port) {
      argv.push('--port', port + '');
    }
    const unwatchs: Function[] = [];
    const isWatch = process.env.WATCH !== 'none';

    function unwatch(): void {
      for (const unwatch of unwatchs) {
        try {
          unwatch();
        } catch (ignored) {}
      }
    }

    function watch(fn: unknown): void {
      if (typeof fn === 'function') {
        unwatchs.push(fn);
      }
    }

    if (absTmpPath) {
      // 保留 '.cache' 和 'node_modules'文件夹
      await clean(absTmpPath, '.cache', 'node_modules');
    }

    watch(await generateFiles({ api, watch: isWatch }));

    const child = fork(require.resolve('@haul-bundler/cli/bin/haul.js'), argv, {
      stdio: 'inherit',
      cwd: absTmpPath,
    });
    child.on('close', (code) => {
      unwatch();
      process.exit(code);
    });

    process.on('SIGINT', () => {
      unwatch();
      child.kill('SIGINT');
    });

    process.on('SIGTERM', () => {
      unwatch();
      child.kill('SIGTERM');
    });
  }

  api.onStart(() => {
    if (
      semver.valid(api.config?.reactNative?.version) &&
      semver.gte(api.config.reactNative.version, '0.59.0') &&
      semver.lt(api.config.reactNative.version, '1.0.0')
    ) {
      return;
    }
    throw new TypeError(`"${name}" 只支持 react-native："0.59.0及以上（>= 0.59.0）" 和 "1.0.0以下（< 1.0.0）" 版本。`);
  });

  api.registerCommand({
    name: 'dev-rn',
    description: 'starts react-native dev webserver',
    fn: handler,
  } as ICommand);

  function detectHaulPresetPath(): string {
    if (semver.valid(api.config?.reactNative?.version)) {
      const minor = semver.minor(api.config.reactNative.version);
      if (minor >= 60) {
        return dirname(require.resolve('@haul-bundler/preset-0.60/package.json'));
      } else {
        return dirname(require.resolve('@haul-bundler/preset-0.59/package.json'));
      }
    }
    throw new TypeError('未知 react-native 版本！');
  }

  api.onGenerateFiles(async () => {
    api.writeTmpFile({
      path: 'babel.config.js',
      content: Mustache.render(babelConfigTpl, {
        presetPath: winPath(dirname(require.resolve('@haul-bundler/babel-preset-react-native/package.json'))),
      }),
    });
    const buildDevUtilsPath = resolve.sync('@umijs/preset-built-in/lib/plugins/commands/buildDevUtils', {
      basedir: process.env.UMI_DIR,
    });
    const { getBundleAndConfigs } = require(buildDevUtilsPath);
    const { bundleConfigs } = await getBundleAndConfigs({ api });
    const config = bundleConfigs.filter((bundleConfig: any) => {
      return bundleConfig.entry?.umi;
    })[0];

    const alias = lodash.omit(config.resolve?.alias, ['react-dom']);
    Object.assign(alias, {
      // haul似乎没有开启treeShaking, 会加载umi common js格式的代码， 导致webpack构建时：out of memory
      umi: resolve.sync('umi/dist/index.esm.js', { basedir: absTmpPath }),
    });
    api.writeTmpFile({
      path: 'haul.config.js',
      content: Mustache.render(haulConfigTpl, {
        haulPresetPath: winPath(detectHaulPresetPath()),
        alias: JSON.stringify(alias, null, 2),
      }),
    });
  });
};
