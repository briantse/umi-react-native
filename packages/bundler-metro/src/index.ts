import { IApi } from '@umijs/types';
import { join } from 'path';
import { existsSync } from 'fs';
import generateFiles from '@umijs/preset-built-in/lib/plugins/commands/generateFiles';
import { cleanTmpPathExceptCache } from '@umijs/preset-built-in/lib/plugins/commands/buildDevUtils';
import { watchPkg } from '@umijs/preset-built-in/lib/plugins/commands/dev/watchPkg';
import loadMetroConfig from '@react-native-community/cli/build/tools/loadMetroConfig';
import loadConfig from '@react-native-community/cli/build/tools/config';
import getIPAddress from './getIPAddress';

export default (api: IApi) => {
  const {
    logger,
    env,
    paths,
    utils: { chalk, portfinder },
  } = api;
  const METRO_PATH = join(api.paths.absNodeModulesPath || process.cwd(), 'metro');

  if (!existsSync(METRO_PATH)) {
    throw new TypeError('未找到 "metro"，请执行 `yarn install` 安装所有依赖。');
  }

  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const MetroServer = require(join(METRO_PATH, 'src', 'Server.js'));
  let port: number;
  let hostname: string;
  let server: typeof MetroServer;
  const unwatchs: (() => void)[] = [];

  function destroy() {
    for (const unwatch of unwatchs) {
      unwatch();
    }
  }

  api.registerCommand({
    name: 'dev-rn',
    description: 'starts react-native dev webserver',
    fn: async ({ args }) => {
      logger.info('rn-dev:', args);
      const defaultPort = process.env.PORT || args?.port || api.config.devServer?.port;
      port = await portfinder.getPortPromise({
        port: defaultPort ? parseInt(String(defaultPort), 10) : 8081,
      });
      hostname = process.env.HOST || api.config.devServer?.host || getIPAddress();
      console.log(chalk.cyan('Starting the development server...'));
      // eslint-disable-next-line no-unused-expressions
      process.send?.({ type: 'UPDATE_PORT', port });

      cleanTmpPathExceptCache({
        absTmpPath: paths.absTmpPath!,
      });
      const watch = process.env.WATCH !== 'none';

      async function runMetroServer(): Promise<void> {
        const ctx = loadConfig(api.paths.cwd);
        const metroConfig = await loadMetroConfig(ctx, {
          ...args,
          watchFolders: [api.paths.absTmpPath || ''],
          projectRoot: api.paths.cwd,
          // reporter,
        });
      }

      function restartMetroServer(): void {
        if (server && typeof server.close === 'function') {
          server.close(() => {
            runMetroServer().then();
          });
        }
      }

      // generate files
      const unwatchGenerateFiles = await generateFiles({ api, watch });
      if (unwatchGenerateFiles) unwatchs.push(unwatchGenerateFiles);
      if (watch) {
        // watch pkg changes
        const unwatchPkg = watchPkg({
          cwd: api.cwd,
          onChange() {
            console.log();
            api.logger.info(`Plugins in package.json changed.`);
            api.restartServer();
          },
        });
        unwatchs.push(unwatchPkg);

        // watch config change
        const unwatchConfig = api.service.configInstance.watch({
          userConfig: api.service.userConfig,
          onChange: async ({ pluginChanged, userConfig, valueChanged }) => {
            if (pluginChanged.length) {
              console.log();
              api.logger.info(`Plugins of ${pluginChanged.map((p) => p.key).join(', ')} changed.`);
              api.restartServer();
            }
            if (valueChanged.length) {
              let reload = false;
              let regenerateTmpFiles = false;
              const fns: Function[] = [];
              const reloadConfigs: string[] = [];
              valueChanged.forEach(({ key, pluginId }) => {
                const { onChange } = api.service.plugins[pluginId].config || {};
                if (onChange === api.ConfigChangeType.regenerateTmpFiles) {
                  regenerateTmpFiles = true;
                }
                if (!onChange || onChange === api.ConfigChangeType.reload) {
                  reload = true;
                  reloadConfigs.push(key);
                }
                if (typeof onChange === 'function') {
                  fns.push(onChange);
                }
              });

              if (reload) {
                console.log();
                api.logger.info(`Config ${reloadConfigs.join(', ')} changed.`);
                api.restartServer();
              } else {
                api.service.userConfig = api.service.configInstance.getUserConfig();

                // TODO: simplify, 和 Service 里的逻辑重复了
                // 需要 Service 露出方法
                const defaultConfig = await api.applyPlugins({
                  key: 'modifyDefaultConfig',
                  type: api.ApplyPluginsType.modify,
                  initialValue: await api.service.configInstance.getDefaultConfig(),
                });
                api.service.config = await api.applyPlugins({
                  key: 'modifyConfig',
                  type: api.ApplyPluginsType.modify,
                  initialValue: api.service.configInstance.getConfig({
                    defaultConfig,
                  }) as any,
                });

                if (regenerateTmpFiles) {
                  await generateFiles({ api });
                } else {
                  fns.forEach((fn) => fn());
                }
              }
            }
          },
        });
        unwatchs.push(unwatchConfig);
      }
    },
  });

  api.registerCommand({
    name: 'build-rn',
    description: 'builds react-native javascript bundle for offline use',
    fn: async ({ args: { platform } }) => {
      logger.info('rn-build:', platform);
      // await generateFiles({ api, watch: false });
    },
  });
};
