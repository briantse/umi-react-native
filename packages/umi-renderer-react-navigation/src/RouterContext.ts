import { Context } from 'react';
import { RouteComponentProps, __RouterContext } from 'react-router';

const context = __RouterContext as Context<RouteComponentProps>;

export default context;
