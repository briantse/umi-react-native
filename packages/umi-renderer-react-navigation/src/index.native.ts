export * from './createHistoryNavigator';
export * from './Navigation';
