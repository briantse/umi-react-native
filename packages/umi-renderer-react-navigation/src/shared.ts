// import { IRoute } from '@umijs/renderer-react';
// import { CommonNavigationAction, NavigationState, StackActionType } from '@react-navigation/routers';
// import { DefaultNavigatorOptions, EventMapBase, DefaultRouterOptions } from '@react-navigation/native';
// import { Action, History, Location, TransitionPromptHook } from 'history';
//
// export type IHistoryNavigationAction = CommonNavigationAction | StackActionType;
//
// export interface IHistoryRouterOptions extends DefaultRouterOptions {
//   history: History<IHistoryNavigationState>;
//   routes: IHistoryScreenOptions[];
// }
//
// export interface IHistoryScreenOptions extends Omit<IRoute, 'component' | 'routes'> {
//   type: 'stack' | 'tab' | 'drawer';
//   [key: string]: unknown; // umi 扩展路由属性
// }
//
// export interface IHistoryNavigationState extends NavigationState {
//   type: 'history';
//   history: Location<IHistoryScreenOptions>[];
// }
//
// export interface IHistoryEventMap extends EventMapBase {
//   historyChange: {
//     data: { location: Location<IHistoryNavigationState>; action: Action };
//     canPreventDefault: false;
//   };
//   historyBlock: { data: { prompt: TransitionPromptHook<IHistoryNavigationState> }; canPreventDefault: true };
// }
//
// export interface IHistoryNavigatorProps extends DefaultNavigatorOptions<IHistoryScreenOptions> {
//   history: History<IHistoryNavigationState>;
// }
