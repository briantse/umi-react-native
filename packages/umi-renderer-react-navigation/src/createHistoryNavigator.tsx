import React from 'react';
import { Platform } from 'react-native';
import {
  createNavigatorFactory,
  useNavigationBuilder,
  StackRouter,
  DefaultNavigatorOptions,
  StackActions,
  EventArg,
} from '@react-navigation/native';
import { StackView } from '@react-navigation/stack';
import { StackRouterOptions, StackNavigationState } from '@react-navigation/routers/lib/typescript/src/StackRouter';
import {
  StackNavigationEventMap,
  StackNavigationConfig,
  StackNavigationOptions,
} from '@react-navigation/stack/lib/typescript/src/types';
import { History, Location, Action } from 'history';
import { parse } from 'querystring';

function HistoryNavigator({
  initialRouteName,
  children,
  screenOptions,
  history,
  ...rest
}: DefaultNavigatorOptions<StackNavigationOptions> &
  StackRouterOptions &
  StackNavigationConfig & { history: History<any> }) {
  const defaultOptions = {
    gestureEnabled: Platform.OS === 'ios',
    animationEnabled: Platform.OS !== 'web',
  };
  const { descriptors, state, navigation } = useNavigationBuilder<
    StackNavigationState,
    StackRouterOptions,
    StackNavigationOptions,
    StackNavigationEventMap
  >(StackRouter, {
    initialRouteName,
    children,
    screenOptions:
      typeof screenOptions === 'function'
        ? (...args) => ({
            ...defaultOptions,
            ...screenOptions(...args),
          })
        : {
            ...defaultOptions,
            ...screenOptions,
          },
  });

  React.useEffect(
    () =>
      navigation.addListener &&
      navigation.addListener('tabPress', (e) => {
        const isFocused = navigation.isFocused();

        // Run the operation in the next frame so we're sure all listeners have been run
        // This is necessary to know if preventDefault() has been called
        requestAnimationFrame(() => {
          if (state.index > 0 && isFocused && !(e as EventArg<'tabPress', true>).defaultPrevented) {
            // When user taps on already focused tab and we're inside the tab,
            // reset the stack to replicate native behaviour
            navigation.dispatch({
              ...StackActions.popToTop(),
              target: state.key,
            });
          }
        });
      }),
    [navigation, state.index, state.key],
  );

  React.useEffect(
    () =>
      history.listen((location: Location<any>, action: Action): void => {
        if (state.routeNames.includes(location.pathname)) {
          switch (action) {
            case 'POP':
              if (navigation.canGoBack()) {
                navigation.dispatch(StackActions.pop());
              }
              break;
            case 'PUSH':
              navigation.dispatch(StackActions.push(location.pathname, parse(location.search.replace('?', ''))));
              break;
            case 'REPLACE':
              navigation.dispatch(StackActions.replace(location.pathname, parse(location.search.replace('?', ''))));
              break;
          }
        }
      }),
    [navigation, history],
  );
  return <StackView {...rest} descriptors={descriptors} state={state} navigation={navigation} />;
}

export const createHistoryNavigator = createNavigatorFactory<
  StackNavigationState,
  StackNavigationOptions,
  StackNavigationEventMap,
  typeof HistoryNavigator
>(HistoryNavigator);
