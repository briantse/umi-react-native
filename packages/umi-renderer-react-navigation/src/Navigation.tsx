import React, { ComponentType, useEffect } from 'react';
import { View, Text } from 'react-native';
import {
  ApplyPluginsType,
  Plugin,
  History,
  Redirect,
  __RouterContext as RouterContext,
  matchPath,
  IRouteProps,
  IRouteComponentProps,
} from 'umi';
import { matchRoutes } from 'react-router-config';
import { Route, NavigationHelpers } from '@react-navigation/native';
import urlJoin from 'url-join';
import md5 from 'blueimp-md5';
import { createHistoryNavigator } from './createHistoryNavigator';

const { Navigator, Screen } = createHistoryNavigator();

interface INavigationProps {
  routes: IRouteProps[];
  plugin: Plugin;
  history: History<any>;
  defaultTitle?: string;
  extraProps?: object;
  [key: string]: any;
}

interface IScreen {
  key: string;
  name: string;
  component: ComponentType<any>;
  options: {
    routeMatchOpts: IRouteProps;
    sensitive?: boolean;
    title?: string;
    [key: string]: any;
  };
}

export interface IScreenComponentProps extends IRouteComponentProps {
  screen: Route<any>;
  navigation: NavigationHelpers<any, any>;
}

function flattenRoutes(routes?: IRouteProps[], parent?: IScreen): IScreen[] {
  if (!Array.isArray(routes)) return [];
  const screens: IScreen[] = [];
  for (let idx = 0; idx < routes.length; idx++) {
    const route = routes[idx];
    const { key: routeKey, path, exact, component, strict, redirect, wrappers, routes: children, ...options } = route;
    const name = parent && parent.name ? urlJoin(parent.name, path || '/') : path || '/';
    const screenKey = routeKey || `s_${md5(name)}`;
    const screen: IScreen = {
      key: screenKey,
      name,
      component: function ScreenComponent(props) {
        if (redirect) {
          return <Redirect from={path} to={redirect} exact={exact} strict={strict} />;
        }
        const children = component ? React.createElement(component, props) : null;
        return parent && parent.component ? React.createElement(parent.component, props, children) : children;
        // if (Array.isArray(wrappers)) {
        //   let len = wrappers.length - 1;
        //   while (len >= 0) {
        //     wrapper = React.createElement(wrappers[len], props, wrapper);
        //     len -= 1;
        //   }
        // }
      },
      options: {
        ...options,
        routeMatchOpts: route,
      },
    };
    if (Array.isArray(children) && children.length > 0) {
      screens.push(...flattenRoutes(children, screen));
    } else {
      screens.push(screen);
    }
  }
  return screens;
}

export function Navigation(props: INavigationProps) {
  const { history, plugin, routes, defaultTitle, extraProps = {}, ...rests } = props;

  useEffect(() => {
    function routeChangeHandler(location: any, action?: string) {
      const matchedRoutes = matchRoutes(props.routes, location.pathname);

      plugin.applyPlugins({
        key: 'onRouteChange',
        type: ApplyPluginsType.event,
        args: {
          routes: props.routes,
          matchedRoutes,
          location,
          action,
        },
      });
    }
    routeChangeHandler(history.location, 'POP');
    return history.listen(routeChangeHandler);
  }, [history]);

  const screens = flattenRoutes(routes);

  if (__DEV__) {
    if (!screens || screens.length === 0) {
      return (
        <Navigator initialRouteName="/" history={history}>
          <Screen
            name="/"
            component={() => (
              <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
                <Text style={{ fontWeight: 'bold', fontSize: 17, color: '#f4333c' }}>404</Text>
                <Text style={{ fontSize: 14, color: '#f4333c' }}>请在 pages/ 目录下实现一个 index 页面。</Text>
              </View>
            )}
          />
        </Navigator>
      );
    }
  }

  return (
    <Navigator initialRouteName="/" history={history}>
      {screens.map(({ key, component: Component, options: { routeMatchOpts, title, ...options }, ...rest }, idx) => (
        <Screen {...rest} key={key || `screen_${idx}`} options={{ ...options, title: title || defaultTitle }}>
          {(props) => {
            const newProps = {
              ...rests,
              ...extraProps,
              ...props,
            };
            return (
              <RouterContext.Provider
                value={{
                  history,
                  location: history.location,
                  match: matchPath(history.location.pathname, routeMatchOpts) || {
                    path: '/',
                    params: {},
                    isExact: true,
                    url: '/',
                  },
                }}
              >
                <Component {...newProps} />
              </RouterContext.Provider>
            );
          }}
        </Screen>
      ))}
    </Navigator>
  );
}
