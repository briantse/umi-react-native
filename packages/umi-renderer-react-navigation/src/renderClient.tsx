import React from 'react';
import { ApplyPluginsType, Plugin } from 'umi';
import { IRoute } from '@umijs/renderer-react';
import { Navigation } from './Navigation';

interface IRouterComponentProps {
  routes: IRoute[];
  plugin: Plugin;
  history: any;
  ssrProps?: object;
  defaultTitle?: string;
}

interface IOpts extends IRouterComponentProps {
  rootElement?: string | HTMLElement;
}

export function renderClient(opts: IOpts) {
  return opts.plugin.applyPlugins({
    type: ApplyPluginsType.modify,
    key: 'rootContainer',
    initialValue: (
      <Navigation history={opts.history} routes={opts.routes} plugin={opts.plugin} defaultTitle={opts.defaultTitle} />
    ),
    args: {
      history: opts.history,
      routes: opts.routes,
      plugin: opts.plugin,
    },
  });
}
