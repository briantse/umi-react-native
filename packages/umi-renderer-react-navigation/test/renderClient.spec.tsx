import { renderClient } from '../src/renderClient';
import { Plugin } from 'umi';
import { render } from '@testing-library/react-native';

jest.mock('../src/Navigation', () => {
  const React = require('react');
  const View = require('react-native').View;
  function Navigation(props: any) {
    return React.createElement(View, props);
  }
  return { Navigation };
});

const plugin = new Plugin({
  validKeys: [
    'patchRoutes',
    'rootContainer',
    'render',
    'onRouteChange',
    'onReactNavigationStateChange',
    'getReactNavigationInitialState',
    'getReactNavigationInitialIndicator',
    'dva',
  ],
});

describe('renderClient', () => {
  it('should work', function () {
    const App = renderClient({
      routes: [],
      plugin,
      history: null,
      defaultTitle: 'title-test',
    });
    const { baseElement } = render(App);
    expect(baseElement).toMatchSnapshot();
  });
});
