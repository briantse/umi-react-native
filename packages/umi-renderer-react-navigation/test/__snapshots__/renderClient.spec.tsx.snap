// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`umi-renderer-react-navigation should renderClient 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      defaultTitle="title-test"
      history={null}
      plugin={
        Plugin {
          "hooks": Object {},
          "validKeys": Array [
            "patchRoutes",
            "rootContainer",
            "render",
            "onRouteChange",
            "onReactNavigationStateChange",
            "getReactNavigationInitialState",
            "getReactNavigationInitialIndicator",
            "dva",
          ],
        }
      }
      routes={Array []}
    />
  </View>
</View>
`;
